### PR TITLE
NemesisIO input

### PR DESCRIPTION
**Pull request by roystgnr**
This adds a basic NemesisIO(DistributedMesh) read() unit test, followed by enough fixes to get that read() working.

I'll be adding more serious testing (and realistically more fixes) in subsequent PRs but this should hopefully be a straightforward improvement already.

However, this shouldn't be merged until we have Civet up and running again - I think the Nemesis CI tests in MOOSE are going to be robust enough to not be fazed by the numbering issue I fixed here without us having to re-gold first, but I won't swear to that.

### CODE REVIEW

File: src/mesh/nemesis_io_helper.C
Comment (jwpeterson) [inline src/mesh/nemesis_io_helper.C:2339]: I haven't thought about this in detail, but my initial reaction is "he's subtracting 1 from a libmesh node id" which seems like it would be bad if the libmesh node id is 0?

Comment (roystgnr) [reply src/mesh/nemesis_io_helper.C:2339]: Sadly, it's not that straightforward.  Subtracting 1 from a libmesh-node-id-plus-one (which gets done because we want 1-based numbering in the file and we directly pass that vector::data() to an Exodus API.  See the "push_back(id+1)" call and preceding comment.

Should we rename that vector?  And move it while we're at it; I'm not sure why it's in exodusII_io_helper.h if it's only ever used by the subclass.

Comment (jwpeterson) [reply src/mesh/nemesis_io_helper.C:2339]: > Should we rename that vector? And move it while we're at it; I'm not sure why it's in exodusII_io_helper.h if it's only ever used by the subclass.

I'd be in favor of both the renaming and moving to the Nemesis helper. Not sure why it was added to the Exodus helper... maybe it was used there at some point.

**Comment by roystgnr** [comment]
In theory (and in my own libMesh testing) this branch now also has nodal data being written and re-read correctly.

In practice, we'll see what the MOOSE tests have to say about that.  If there's any regressions there then I'm going to regret not letting the original branch go through CI before I tacked on dozens more commits...

Since I've had to futz quite a bit with the heavily-used ExodusII_IO_Helper class here, and still I'm not yet familiar enough nor happy enough with that code, I'm leaning towards not including this PR in our next release branch, even if I'm lucky and CI can't find any problems with it.

**Comment by roystgnr** [comment]
Those unit tests are passing for me on 1 through 16 processors now.  Hopefully --enable-complex works too.  I'll bet I don't have my ifdefs for --disable-nemesis/--disable-exodus in perfect order, so I'll check on that before we merge, but otherwise I think this is at a good break point.

**Review by jwpeterson — APPROVED** [review]
Looks good to me assuming you can figure out what's going on in the failing complex-variables build, but it's quite large (GitHub refused to even display the diffs in nemesis_io.C and nemesis_io_helper.C) so I probably haven't given it the most thorough possible review...

**Comment by roystgnr** [comment]
I don't have enough hubris to say I've replicated **the** problem, but I can trigger **a** problem in two of the new unit tests in a complex build, so I'll see if I can fix those and we'll let CI take another crack at it from there.

**Comment by roystgnr** [comment]
If you do want to review those huge diffs, going commit by commit might work.  I've been trying to keep each individual edit as small and self-contained as possible, and I've already tried to wipe out any blind alleys with fixup commits so hopefully nothing left in the intermediate commits is redundant/obsoleted either.

**Comment by roystgnr** [comment]
The complex case was a pretty clean fix.  I'll merge tomorrow morning unless anyone screams.

**Pull request merged**
